cmake: Some improvements using `PROJECT_IS_TOP_LEVEL` variable

hebasto opened the pull request:
This PR:
1. Emulates [`PROJECT_IS_TOP_LEVEL`](https://cmake.org/cmake/help/latest/variable/PROJECT_IS_TOP_LEVEL.html) variable for CMake versions where it is not available.
2. Makes the `SECP256K1_INSTALL` option dependent on `PROJECT_IS_TOP_LEVEL` (a [follow up](https://github.com/bitcoin-core/secp256k1/pull/1263#issuecomment-1516564300) of https://github.com/bitcoin-core/secp256k1/pull/1263).
3. Makes integration of this project as a subtree easier. A top project can `#include <secp256k1.h>` with no additional `target_include_directories()` commands. For example, see https://github.com/hebasto/secp256k1-CMake-example/tree/subtree.

real-or-random commented on CMakeLists.txt:
```diff
@@ -26,13 +38,14 @@ set(CMAKE_C_EXTENSIONS OFF)
 endif()
 
-option(SECP256K1_INSTALL "Enable installation" ON)
+include(CMakeDependentOption)
+cmake_dependent_option(SECP256K1_INSTALL "Enable installation." ON "PROJECT_IS_TOP_LEVEL" OFF)
```
Does it really make sense to forbid installation if used as a subproject? I think we could let the user decide here.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/secp256k1/pull/1284#issuecomment-1517661110).

real-or-random reviewed (commented):
Concept ACK

hebasto commented:
Updated b9112aa61b9a0cd9e6f790b920568b85be1d5b8a -> 009faea5c6f48b0d7b7c2bf87caf44621b6c6b21 ([pr1284.01](https://github.com/hebasto/secp256k1/commits/pr1284.01) -> [pr1284.02](https://github.com/hebasto/secp256k1/commits/pr1284.02), [diff](https://github.com/hebasto/secp256k1/compare/pr1284.01..pr1284.02)):

- addressed @real-or-random's [comment](https://github.com/bitcoin-core/secp256k1/pull/1284#discussion_r1173480510)

real-or-random reviewed (approved):
utACK 009faea5c6f48b0d7b7c2bf87caf44621b6c6b21

real-or-random reviewed (approved):
[edited: nevermind, I acked the wrong PR...]

theuni commented on src/CMakeLists.txt:
```diff
@@ -31,7 +31,8 @@ endif()
 
-target_include_directories(secp256k1 PUBLIC
+target_include_directories(secp256k1 INTERFACE
+  $<BUILD_INTERFACE:$<$<NOT:$<BOOL:${PROJECT_IS_TOP_LEVEL}>>:${PROJECT_SOURCE_DIR}/include>>
```
Can we get a comment? Something like:
"Add the include path for parent projects so that they don't have to manually add it."

theuni commented:
Hah, that was fast!

utACK 71f746c057a66d5ae0dfdb5d439bc49592c0d16e.
